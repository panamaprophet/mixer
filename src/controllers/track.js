'use strict'

import FaderController from './fader'
import TrackTemplate from '../views/track'

class TrackController {

	constructor(options) {

		this.track = options.track
		this.mountPoint = options.mountPoint

		this.template = document.createElement('template')
		this.template.innerHTML = TrackTemplate


		this.el = this.layout()
	}

	layout() {

		let track = this.track
		let element = document.importNode(this.template.content, true)


		element.querySelector('[data-name="title"]').textContent = track.title


		element.querySelector('[data-name="mute"]').addEventListener('click', e => {
			track.toggleMute()
			e.currentTarget.classList.toggle('button--pressed', track.muted)
		})


		element.querySelector('[data-name="bypass"]').addEventListener('click', e => {
			track.toggleFX()
			e.currentTarget.classList.toggle('button--pressed', track.bypassFX)
		})



		let fader = element.querySelector('[data-name="fader"]')

		let faderView = new FaderController({
			min: 0,
			max: 1,
			step: 0.05,
			value: track.ready ? 1 : 0,
			vertical: true,
			onChange: function(value) {
				track.volume = value
			}
		}, fader)


		if (!track.ready) {

			track.onReady = function(track) {
				faderView.animateTo(1)
			}

		}



		let faders = element.querySelectorAll('[data-type="fader"]')

		for (let i = 0; i < faders.length; i++) {

			let fxIdent = faders[i].getAttribute('data-fx')

			let fader = new FaderController({
				min: 0,
				max: 0.95,
				value: 0,
				step: 0.05,
				onChange: function(value) {
					track.fx[fxIdent].gain.value = value
				}
			}, faders[i])
		}



		if (this.mountPoint) {
			this.mountPoint.appendChild(element)
		}

		return element
	}
}


export default TrackController