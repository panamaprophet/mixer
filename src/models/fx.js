import Delay from './fx/delay'
import Reverb from './fx/reverb'
import Distortion from './fx/distortion'

export { Delay, Reverb, Distortion }