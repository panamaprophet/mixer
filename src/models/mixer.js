'use strict'

import {map} from 'ramda';


import Track from './track';

import {
	Delay,
	Reverb,
	Distortion,
} from './fx';

import {
	createContext,
	createAnalyser,
	createMasterBus,
	createTrackFromSource,
} from '/helpers/audio';

import {
	playAll,
	pauseAll,
	rewindAll,
} from '/helpers/playback';


class Mixer {
	constructor(sources, onReady) {
		this.context = createContext();
		this.analyser = createAnalyser(this.context);
		this.masterBus = createMasterBus(this.context, [this.analyser]);
		
		this.tracks = [];

		this.fx = [
			new Delay(this.context, this.masterBus),
			new Reverb(this.context, this.masterBus),
			new Distortion(this.context, this.masterBus),
		];

		const trackPromises = map(createTrackFromSource(this.context, this.masterBus), sources);

		Promise.all(trackPromises)
			.then(map(track => track.addFx(this.fx)))
			.then(map(track => this.tracks.push(track)))
			.then(map(onReady));
	}

	play() {
		return playAll(this.tracks);
	}

	pause() {
		return pauseAll(this.tracks);
	}

	rewind() {
		return rewindAll(this.tracks);
	}
}


export default Mixer;