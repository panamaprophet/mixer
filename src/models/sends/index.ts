import type {Delay} from './delay';
import type {Reverb} from './reverb';
import type {Distortion} from './distortion';

export * from './reverb';
export * from './delay';
export * from './distortion';


export type SendId = string;

export type SendParamValue = number | string | Float32Array | AudioBuffer | BiquadFilterType;

export type Send = Reverb | Delay | Distortion;

export type SendConstructor = new (context: AudioContext, masterBus: GainNode) => Send;