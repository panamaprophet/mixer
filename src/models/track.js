'use strict';

import {fetchAudioAsArrayBuffer} from '/helpers/audio';
import {
    connectNodes,
    createGainNode,
    getNodeParamNormalizedValue,
    setNodeParamNormalizedValue,
} from '/helpers/node';


const generateIdByTitle = title => title.replace(/[^A-Za-z]+/gi, '').toLowerCase();


class Track {
    constructor({url, title, context, masterBus}) {
        this.id = generateIdByTitle(title);

        this.source = null;
        this.buffer = null;

        this.title = title;
        this.context = context;

        this.pausedAt = 0;
        this.startedAt = 0;

        this.muted = false;
        this.playing = false;
        this.ready = false;
        this.bypassFX = false;

        this.fx = {};

        this.bus = createGainNode(context);
        connectNodes(this.bus, masterBus);

        this.loadingState = this.load(url);
    }


    get volume() {
        return getNodeParamNormalizedValue(this.bus.gain);
    }

    set volume(value) {
        if (this.muted) {
            this.previousVolume = value;
        } else {
            setNodeParamNormalizedValue(this.bus.gain, value);
        }
    }

    load(url) {
        return fetchAudioAsArrayBuffer(url)
            .then(audioBuffer => this.context.decodeAudioData(audioBuffer))
            .then(decodedAudioData => {
                this.buffer = decodedAudioData;
                this.ready = true;

                console.log('Track "%s" is ready', this.title);

                return this;
            })
            .catch(error => console.log('[ERROR LOADING TRACK]', error));
    }


    play() {
        if (this.playing) {
            return false;
        }

        this.source = this.context.createBufferSource()
        this.source.buffer = this.buffer
        this.source.connect(this.bus)
        this.source.start(0, this.pausedAt)

        this.startedAt = this.context.currentTime - this.pausedAt
        this.pausedAt = 0
        this.playing = true
    }

    pause() {
        const elapsed = this.context.currentTime - this.startedAt;
        this.stop();
        this.pausedAt = elapsed;
    }

    stop() {
        if (this.source) {
            this.source.disconnect()
            this.source.stop(0)
            this.source = null
        }

        this.pausedAt = 0
        this.startedAt = 0
        this.playing = false
    }

    mute() {
        this.previousVolume = this.volume;
        this.volume = 0;
        this.muted = true;
    }

    unmute() {
        this.muted = false;
        this.volume = this.previousVolume;
    }

    toggleMute() {
        this.muted ? this.unmute() : this.mute();
    }

    addFx(effects) {
        return effects.map(fx => {
            const {id, signalIn} = fx;
            const bus = createGainNode(this.context, 0);

            if (this.fx[id]) {
                return false;
            }

            connectNodes(this.bus, bus);
            connectNodes(bus, signalIn);

            this.fx[id] = bus;

            return bus;
        });
    }

    removeFx(id) {
        this.fx[id].disconnect();
    }

    toggleFX() {
        const fxNames = Object.keys(this.fx)

        if (this.bypassFX) {

            fxNames.forEach(fxName => {
                this.fx[fxName].gain.value = this.fx[fxName].previousVolume
            })

        } else {

            fxNames.forEach(fxName => {
                this.fx[fxName].previousVolume = this.fx[fxName].gain.value
                this.fx[fxName].gain.value = 0
            })

        }

        this.bypassFX = !this.bypassFX
    }
}


export default Track