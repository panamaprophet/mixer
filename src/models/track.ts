import {createPanner, fetchAudioAsArrayBuffer} from '/helpers/audio';
import {generateIdByTitle} from '/helpers/entities';
import {connectNodes, createGainNode, getNodeParamNormalizedValue, setNodeParamNormalizedValue} from '/helpers/node';
import type {SendId, Send} from './sends/index';


export type TrackId = string;

export enum TrackState {
    NOT_SET,
    LOADING,
    READY,
    FAILED,
}

export type TrackSource = {
    url: string,
    title: string,
}

export interface TrackInterface {
    id: TrackId,
    title: string,
    sends: Record<SendId, GainNode>,
    isMuted: boolean,
    isPlaying: boolean,
    isSendsEnabled: boolean,
    state: TrackState,
    loadPromise: Promise<TrackState>,
    addSend: (send: Send) => GainNode | null,
    load: (url: string) => void,
    play: () => void,
    pause: () => void,
    stop: () => void,
    mute: () => void,
    unmute: () => void,
    toggleMute: () => boolean,
    setSendLevel(id: SendId, value: number): void,
}

type TrackProps = {
    url: string,
    title: string,
    context: AudioContext,
    masterBus: GainNode,
    sends: Send[],
}

export class Track implements TrackInterface {
    id: TrackId;
    title: string;

    #context: AudioContext;
    #panner: StereoPannerNode;
    #bus: GainNode;
    #source: AudioBufferSourceNode | null = null;
    #buffer: AudioBuffer | null = null;

    sends: Record<SendId, GainNode> = {};
    #sendLevels: Record<SendId, number> = {};

    #pausedAt = 0;
    #startedAt = 0;
    #previousVolume = 0;

    isMuted = false;
    isPlaying = false;
    isSendsEnabled = true;

    state = TrackState.NOT_SET;
    loadPromise: Promise<TrackState> = Promise.resolve(this.state);


    constructor({url, title, context, masterBus, sends = []}: TrackProps) {
        this.id = generateIdByTitle(title);
        this.title = title;
        this.#context = context;
        this.#bus = createGainNode(context);
        this.#panner = createPanner(context);

        connectNodes(this.#bus, this.#panner);
        connectNodes(this.#panner, masterBus);

        sends.map(send => this.addSend(send));

        void this.load(url);
    }

    get volume(): number {
        return getNodeParamNormalizedValue(this.#bus.gain);
    }

    set volume(value: number) {
        if (this.isMuted) {
            this.#previousVolume = value;
        } else {
            setNodeParamNormalizedValue(this.#bus.gain, value);
        }
    }

    get pan(): number {
        return getNodeParamNormalizedValue(this.#panner.pan);
    }

    set pan(value: number) {
        setNodeParamNormalizedValue(this.#panner.pan, value);
    }

    addSend({id, signalIn}: Send): GainNode | null {
        if (this.sends[id]) {
            return null;
        }

        const bus = createGainNode(this.#context, 0);

        connectNodes(this.#bus, bus);
        connectNodes(bus, signalIn);

        this.sends[id] = bus;

        return bus;
    }

    removeSend(id: SendId): void {
        const send = this.sends[id];

        send?.disconnect();
    }

    toggleSends(): boolean {
        const sendIds = Object.keys(this.sends);

        sendIds.forEach(sendId => {
            const send = this.sends[sendId];

            if  (this.isSendsEnabled) {
                this.#sendLevels[sendId] = getNodeParamNormalizedValue(send.gain);
                setNodeParamNormalizedValue(send.gain, 0);
            } else {
                setNodeParamNormalizedValue(send.gain, this.#sendLevels[sendId]);
            }
        });

        this.isSendsEnabled = !this.isSendsEnabled;

        return this.isSendsEnabled;
    }

    async load(url: string): Promise<TrackState> {
        this.state = TrackState.LOADING;

        this.loadPromise = new Promise((resolve) => {
            fetchAudioAsArrayBuffer(url)
                .then(arrayBuffer => this.#context.decodeAudioData(arrayBuffer))
                .then(audioBuffer => {
                    this.#buffer = audioBuffer;
                    this.state = TrackState.READY;
                })
                .catch(error => {
                    this.state = TrackState.FAILED;
                    console.log('[ERROR LOADING TRACK]', error);
                })
                .finally(() => {
                    resolve(this.state);
                });
            });

        return this.loadPromise;
    }

    play(): void {
        if (this.isPlaying) {
            return;
        }

        this.#source = this.#context.createBufferSource();
        this.#source.buffer = this.#buffer;
        this.#source.connect(this.#bus);
        this.#source.start(0, this.#pausedAt);

        this.#startedAt = this.#context.currentTime - this.#pausedAt;
        this.#pausedAt = 0;
        this.isPlaying = true;
    }

    pause(): void {
        if (this.isPlaying) {
            const elapsed = this.#context.currentTime - this.#startedAt;
            this.stop();
            this.#pausedAt = elapsed;
        }
    }

    stop(): void {
        if (this.#source) {
            this.#source.disconnect();
            this.#source.stop(0);
            this.#source = null;
        }

        this.#pausedAt = 0;
        this.#startedAt = 0;
        this.isPlaying = false;
    }


    mute(): void {
        this.#previousVolume = this.volume;
        this.volume = 0;
        this.isMuted = true;
    }

    unmute(): void {
        this.isMuted = false;
        this.volume = this.#previousVolume;
    }

    toggleMute(): boolean {
        if (this.isMuted) {
            this.unmute();
        } else {
            this.mute();
        }

        return this.isMuted;
    }

    setSendLevel(id: SendId, value: number): void {
        const send = this.sends[id];

        setNodeParamNormalizedValue(send.gain, value);
    }
}