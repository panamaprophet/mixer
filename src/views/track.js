'use strict'


let Template = `<div class="fader">
					<div data-name="fader" class="fader__placeholder"></div>
					<div class="fader__buttons">
						<button class="fader__button button" data-name="mute">Mute</button>
						<button class="fader__button button" data-name="bypass">Bypass FX</button>
					</div>
					<div class="fader__effects">
						<div class="fader__effect">
							<span class="fader__effect-title">Delay:</span>
							<div class="fader__effect-placeholder" data-type="fader" data-fx="delay" data-max="0.95" data-step="0.05"></div>
						</div>
						<div class="fader__effect">
							<span class="fader__effect-title">Revrb:</span>
							<div class="fader__effect-placeholder" data-type="fader" data-fx="reverb" data-max="0.95" data-step="0.05"></div>
						</div>
						<div class="fader__effect">
							<span class="fader__effect-title">Distrt:</span>
							<div class="fader__effect-placeholder" data-type="fader" data-fx="distortion" data-max="0.95" data-step="0.05"></div>
						</div>
					</div>
					<div class="fader__title" data-name="title"></div>
				</div>`


export default Template