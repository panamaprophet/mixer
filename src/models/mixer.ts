import {createContext, createAnalyser, createMasterBus, isContextRunning, resumeContext} from '/helpers/audio';
import {playAll, pauseAll, rewindAll, stopAll} from '/helpers/playback';
import {Track} from './track';
import type {TrackId, TrackSource} from './track';
import type {SendId, SendParamValue, Send, SendConstructor} from './sends/index';


interface MixerInterface {
    analyser: AnalyserNode;
}


export class Mixer implements MixerInterface {
    #context: AudioContext;
    #masterBus: GainNode;
    analyser: AnalyserNode;
    sends: Send[] = [];
    tracks: Track[] = [];

    constructor(sources: TrackSource[] = [], sends: SendConstructor[] = []) {
        this.#context = createContext();
        this.analyser = createAnalyser(this.#context);
        this.#masterBus = createMasterBus(this.#context, [this.analyser]);

        sends.forEach(send => {
            this.sends.push(new send(this.#context, this.#masterBus));
        });

        void this.load(sources);
    }

    /**
     * @returns {Promise<Mixer>}
     */
    async play(): Promise<this> {
        if (isContextRunning(this.#context) === false) {
            await resumeContext(this.#context);
        }

        playAll(this.tracks);

        return Promise.resolve(this);
    }

    /**
     * @returns {Promise<Mixer>}
     */
    pause(): Promise<this> {
        pauseAll(this.tracks);

        return Promise.resolve(this);
    }

    /**
     * @returns {Promise<Mixer>}
     */
    rewind(): Promise<this> {
        rewindAll(this.tracks);

        return Promise.resolve(this);
    }

    /**
     * @returns {Promise<Mixer>}
     */
    stop(): Promise<this> {
        stopAll(this.tracks);

        return Promise.resolve(this);
    }

    /**
     * @param {TrackId} id
     * @param {number} volume
     * @returns {Promise<Track[]>}
     */
    setTrackVolume(id: TrackId, volume: number): Promise<this> {
        this.tracks.map(track => {
            if (track.id === id) {
                track.volume = volume;
            }

            return track;
        });

        return Promise.resolve(this);
    }

    /**
     *
     * @param {TrackId} id
     * @param {SendId} sendId
     * @param {number} level
     * @returns {Promise<Track[]>}
     */
    setTrackSendLevel(id: TrackId, sendId: SendId, level: number): Promise<this> {
        this.tracks.map(track => {
            if (track.id === id) {
                track.setSendLevel(sendId, level);
            }

            return track;
        });

        return Promise.resolve(this);
    }

    /**
     * @param {TrackId} id
     * @param {number} value
     * @returns {Promise<Track[]>}
     */
    setTrackPanLevel(id: TrackId, value: number): Promise<this> {
        this.tracks.map(track => {
            if (track.id === id) {
                track.pan = value;
            }

            return track;
        });

        return Promise.resolve(this);
    }

    /**
     * @param {TrackId} id
     * @returns {Promise<Track[]>}
     */
    toggleTrack(id: TrackId): Promise<this> {
        this.tracks.map(track => {
            if (track.id === id) {
                track.toggleMute();
            }

            return track;
        });

        return Promise.resolve(this);
    }

    /**
     * @param {TrackId} id
     * @returns {Promise<Track[]>}
     */
    toggleTrackFx(id: TrackId): Promise<this> {
        this.tracks.map(track => {
            if (track.id === id) {
                track.toggleSends();
            }

            return track;
        });

        return Promise.resolve(this);
    }

    /**
     *
     * @param {SendId} sendId
     * @param {number|string} value
     * @retruns {Promise<Send[]>}
     */
    setSendParamValue(sendId: SendId, parameterId: string, value: SendParamValue): Promise<this> {
        this.sends.map(send => {
            if (send.id === sendId) {
                send[parameterId] = value;
            }

            return send;
        });

        return Promise.resolve(this);
    }

    load(sources: TrackSource[]): Promise<this> {
        this.tracks = sources.map(({url, title}) => new Track({
            url,
            title,
            context: this.#context,
            masterBus: this.#masterBus,
            sends: this.sends,
        }));

        const trackLoadingPromises = this.tracks.map(({loadPromise}) => loadPromise);

        return Promise.all(trackLoadingPromises).then(() => this);
    }
}