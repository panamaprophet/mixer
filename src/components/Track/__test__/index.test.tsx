import React from 'react';
import {shallow, mount} from 'enzyme';
import {Track} from '../';
import {TrackState} from '/models/track';


const propsMock = {
    id: 'track',
    title: 'Track',
    volume: 42.0,
    isMuted: false,
    isSendsEnabled: true,
    send: {
        delay: 0,
        reverb: 0,
    },
    state: TrackState.NOT_SET,
    onMute: jest.fn(),
    onBypass: jest.fn(),
    onVolumeChange: jest.fn(),
    onSendLevelChange: jest.fn(() => jest.fn()),
};

const MUTE_BUTTON = 0;
const FX_BYPASS_BUTTON = 1;
const FADER_VOLUME = 0;
const FADER_SEND = 1;


describe('<Track />', () => {
    it('should renders without any errors', () => {
        const wrapper = shallow(
            <Track {...propsMock} />
        );

        expect(wrapper).toMatchSnapshot();
    });

    describe('onMute', () => {
        it('should be called on mute toggler click', () => {
            const wrapper = shallow(
                <Track {...propsMock} />
            );

            wrapper.find('button').at(MUTE_BUTTON).simulate('click');

            expect(propsMock.onMute).toBeCalled();
        });
    });

    describe('onBypass', () => {
        it('should be called on fx bypass toggler click', () => {
            const wrapper = shallow(
                <Track {...propsMock} />
            );

            wrapper.find('button').at(FX_BYPASS_BUTTON).simulate('click');

            expect(propsMock.onBypass).toBeCalled();
        });
    });

    describe('onVolumeChange', () => {
        it('should be called on volume change', () => {
            const wrapper = mount(
                <Track {...propsMock} />
            );

            wrapper.find('Fader').at(FADER_VOLUME).find('FaderThumb').simulate('mouseDown');

            window.document.documentElement.dispatchEvent(new Event('mousemove'));
            window.document.documentElement.dispatchEvent(new Event('mouseup'));

            expect(propsMock.onVolumeChange).toBeCalled();
        });
    });

    describe('onSendLevelChange', () => {
        it('should be called on send level change', () => {
            const wrapper = mount(
                <Track {...propsMock} />
            );

            wrapper.find('Fader').at(FADER_SEND).find('FaderThumb').simulate('mouseDown');

            window.document.documentElement.dispatchEvent(new Event('mousemove'));
            window.document.documentElement.dispatchEvent(new Event('mouseup'));

            expect(propsMock.onSendLevelChange).toBeCalled();
        });
    });
});