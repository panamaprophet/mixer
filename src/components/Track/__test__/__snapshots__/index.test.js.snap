// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Track /> should renders without any errors 1`] = `
<div
  className="track"
>
  <Fader
    isVertical={true}
    onChange={[MockFunction]}
    value={42}
  />
  <div
    className="buttons"
  >
    <button
      className="button"
      onClick={[Function]}
    >
      Mute
    </button>
    <button
      className="button"
      onClick={[Function]}
    >
      Bypass FX
    </button>
  </div>
  <div
    className="sends"
  >
    <div
      className="send"
      key="delay"
    >
      <span
        className="sendTitle"
      >
        delay
      </span>
      <Fader
        value={0}
      />
    </div>
    <div
      className="send"
      key="reverb"
    >
      <span
        className="sendTitle"
      >
        reverb
      </span>
      <Fader
        value={0}
      />
    </div>
  </div>
  <div
    className="title"
  >
    Track
  </div>
</div>
`;
